### PR TITLE
sql: type check `where` clauses

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

File: sql/scan.go [hunk 1]
```diff
@@ -337,7 +337,16 @@ func (n *scanNode) initWhere(where *parser.Where) error {
...
 		// Normalize the expression (this will also evaluate any branches that are
 		// constant).
-		n.filter, n.err = parser.NormalizeAndTypeCheckExpr(n.filter)
+		n.filter, n.err = parser.NormalizeExpr(n.filter)
+	}
+	var whereType parser.Datum
+	if n.err == nil {
+		whereType, n.err = parser.TypeCheckExpr(n.filter)
+	}
+	if n.err == nil {
```
Comment (petermattis) [inline sql/scan.go]: Is there a reason you're dropping out of the if-branch where `whereType` is set in order to check it again? I would have thought this should look like:

```
  if n.err == nil {
    var whereType parser.Datum
    if whereType, n.err = parser.TypeCheckExpr(n.fitler); n.err == nil && whereType != parser.DummyBool {
      n.err = ...
    }
  }
```

File: sql/scan.go [hunk 2]
```diff
@@ -603,12 +612,7 @@ func (n *scanNode) filterRow() bool {
 		return false
 	}
 
-	v, ok := d.(parser.DBool)
-	if !ok {
-		n.err = fmt.Errorf("WHERE clause did not evaluate to a boolean")
-		return false
-	}
-	return bool(v)
+	return bool(d.(parser.DBool))
```
Comment (petermattis) [inline sql/scan.go]: What happens if the WHERE clause evaluates to `NULL`? I think you should add a test where the WHERE clause is something like `a = NULL`.

File: sql/testdata/select [hunk 3]
```diff
@@ -16,6 +16,10 @@ two four
 statement ok
 CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT)
 
+query error argument of WHERE must be type bool, not type string
+SELECT * FROM abc where 'hello'
```
Comment (petermattis) [inline sql/testdata/select]: `s/where/WHERE/g`

File: sql/scan.go (hunk 1 above)
Comment (petermattis) [inline sql/scan.go]: Per my other comment, I think you might need to allow `whereType == parser.DNull` as well.

**Comment by petermattis** [comment]
LGTM modulo handling of WHERE clauses which evaluate to `NULL`.

File: sql/scan.go (hunk 1 above)
Comment (tamird) [inline sql/scan.go]: checked in postgres and fixed

**Comment by petermattis** [comment]
LGTM

**Pull request merged**